### PR TITLE
8324067: GenShen: Isolate regulator thread to generational mode

### PR DESCRIPTION
**Pull request by earthling-amzn**
Shenandoah's generational mode uses a second thread to evaluate heuristics. This is necessary so that the heuristics may interrupt a control thread which is running an old cycle in order to run a young cycle.

The changes here move the regulator thread into `ShenandoahGenerationalHeap`. The generational version of the control thread is also now instantiated only by the generational heap. The upstream version of the control thread has more or less been restored. To  summarize:
* An abstract base class called `ShenandoahController` has been introduced as the base class for the original and generational control threads. It has just one virtual method and it is not on a fast path. Much of the common code has been pulled up into this class.
* The respective control threads no longer need to check what mode they are in. They also no longer need to select which global generation they need to use.  The regulator thread is now only used by the generational mode so it no longer supports running only global cycles.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [JDK-8324067](https://bugs.openjdk.org/browse/JDK-8324067): GenShen: Isolate regulator thread to generational mode (**Task** - P4)

### Reviewers
 * [Kelvin Nilsen](https://openjdk.org/census#kdnilsen) (@kdnilsen - Committer) ⚠️ Review applies to [fabeb853](https://git.openjdk.org/shenandoah/pull/391/files/fabeb853816987f26bf5d3c2f63f862d51edb889)
 * [Y. Srinivas Ramakrishna](https://openjdk.org/census#ysr) (@ysramakrishna - Committer) ⚠️ Review applies to [fabeb853](https://git.openjdk.org/shenandoah/pull/391/files/fabeb853816987f26bf5d3c2f63f862d51edb889)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/shenandoah.git pull/391/head:pull/391` \
`$ git checkout pull/391`

Update a local copy of the PR: \
`$ git checkout pull/391` \
`$ git pull https://git.openjdk.org/shenandoah.git pull/391/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 391`

View PR using the GUI difftool: \
`$ git pr show -t 391`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/shenandoah/pull/391.diff">https://git.openjdk.org/shenandoah/pull/391.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/shenandoah/pull/391#issuecomment-1936515437)

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp:1]: This file is largely reverted to the upstream's version.

File: src/hotspot/share/gc/shenandoah/shenandoahGenerationalControlThread.cpp
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/shenandoahGenerationalControlThread.cpp:1]: The genshen version of shenandoahControlThread.cpp has been renamed to shenandoahGenerationalControlThread.cpp.

**Review by kdnilsen — APPROVED** [review]
Thanks for separating this out.

**Review by ysramakrishna — APPROVED** [review]
This looks very good; left a minor comment. Not necessarily actionable here or now, but just something to think about.

File: src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp [hunk 1]
```diff
@@ -29,83 +28,62 @@
...
+  _requested_gc_cause(GCCause::_no_cause_specified),
+  _degen_point(ShenandoahGC::_degenerated_outside_cycle) {
   set_name("Shenandoah Control Thread");
-  reset_gc_id();
   create_and_start();
 }
 
 void ShenandoahControlThread::run_service() {
-  ShenandoahHeap* const heap = ShenandoahHeap::heap();
+  ShenandoahHeap* heap = ShenandoahHeap::heap();
```
[the remark this answers is not in the document]
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp:53]: Yes, I made most of these declarations `const` here: https://github.com/openjdk/jdk/pull/17795/files. Though for some reason, I missed the `heap` variable so I'll fix that before integrating.

File: src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp [hunk 2]
```diff
@@ -335,7 +257,7 @@ void ShenandoahControlThread::run_service() {
         heap->pacer()->setup_for_idle();
       }
     } else {
-      // Allow pacer to know we have seen this many allocations
+      // Allow allocators to know we have seen this much regions
```
[the remark this answers is not in the document]
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp:260]: Yes, I'll restore that change.

**Pull request closed without merging**